### PR TITLE
browser file

### PR DESCRIPTION
**Pull request by sionking**
* NO LOGIC HAVE CHANGED
This will simply move *some* browser related methods to a separate file, so it the future we can make a browser class that will be modulated out from instapy class.

Changes:
-- browser file
-- don't sig process if threaded
-- proxy fix in settings

This is not the way I wanted to end up with browser handling but it is one step forward to making a class and handlers for browser not to be part of instapy.

### CODE REVIEW

**Comment by sionking** [comment]
@timgrossmann is that change welcome ?
I think it should be tested by some users to see no configuration misbehave.

File: instapy/browser.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+    # TODO: need to move to called
+    message = "Session started!"
+    highlight_print('browser', message, "initialization", "info", logger)
+    print('')
+
+    return browser, err_msg
+
+
+def set_selenium_remote_session(self,
```
Comment (uluQulu) [inline instapy/browser.py]: `self` even there is no any class?

File: instapy/instapy.py [hunk 2]
```diff
@@ -3840,27 +3741,9 @@ def pick_mutual_following(self,
...
-            # close web browser
-            try:
-                self.browser.quit()
-            except Exception as exc:
-                if isinstance(exc, WebDriverException):
-                    self.logger.exception(
-                        "Error occurred while "
-                        "closing web browser!\n\t{}"
-                        .format(str(exc).encode("utf-8")))
+        close_browser(self.browser, False, self.logger)
```
Comment (uluQulu) [inline instapy/instapy.py:3768]: Make clearing cookies a separate function?

File: instapy/util.py [hunk 3]
```diff
@@ -872,18 +872,20 @@ def web_address_navigator(browser, link):
 
 
 @contextmanager
-def interruption_handler(SIG_type=signal.SIGINT, handler=signal.SIG_IGN, notify=None, logger=None):
+def interruption_handler(threaded=False, SIG_type=signal.SIGINT, handler=signal.SIG_IGN, notify=None, logger=None):
```
Comment (uluQulu) [inline instapy/util.py:875]: hm 🤔
althought it won't be a problem [for now]

**Comment by uluQulu** [comment]
Good job, @sionking 
Moving it to a separate file is good 🤝🏼

File: instapy/util.py (hunk 3 above)
Comment (sionking) [reply instapy/util.py:875]: It did for me :)

File: instapy/instapy.py (hunk 2 above)
Comment (sionking) [reply instapy/instapy.py:3768]: all browser related ... why is that a problem ?

File: instapy/browser.py (hunk 1 above)
Comment (sionking) [reply instapy/browser.py]: correct I just did not replace this remote session yet... Will fix!

**Comment by timgrossmann** [comment]
@sionking Thank you for the quick adaption of this idea!
I'll take some time tomorrow to do a review of the code like @uluQulu did, even though I think that he did a great job in it.

I totally agree on the testing thing, let me just paste this list of people @uluQulu once created here 😄 
@fgisslen @jeremycjang @tompicca @lucanello @meinert @cormo1990 @aku6640 @zsvendo @the-unknown @redsector72 @bianpratama @Tachenz @aleeepp @johnvfitness @marcomokastyle @torbenberger @ArdySan2012 @Mehran @GabrieleCalarota @dbmathis @HermanoCrespo @loopypanda @Nocturnal-2 @darokif @Edhim @MMorte @sevilyilmaz @DimIsaev @SkylimiTY @ortegafernando @FranklinClinton @Wisdomlocker @riccardorighettiphoto @DanielSemmelrock @underclass @samille @clevebounce

Would be perfect if some of you guys could give it a shot and report any issues you're facing.
Thank you

**Comment by Mehran** [comment]
can someone explain what is exactly do?

File: instapy/instapy.py (hunk 2 above)
Comment (timgrossmann) [reply instapy/instapy.py:3768]: I think doing this separation of concerns with the separate browser class, we should bundle everything that is part of the browser and put it into the browser file, ending the session included.

Thanks guys for the passionate work!

**Comment by timgrossmann** [comment]
@Mehran Sure, I hope it's the correct explanation. @sionking please correct me if I'm wrong 😉 

This PR starts to separate the logic for InstaPy from the logic of setting up/handling the browser session initiated with selenium. 
Looking at the issues section, we can see that there are several problems regarding selenium and the session creation/handling/ending.
By separating the logic for selenium from the rest of the code, we can introduce better error handling into the browser file itself, keeping it separated from the rest of the InstaPy logic. 
In the end, this will allow for a more stable InstaPy.

This PR starts to introduce the first steps by pulling out (refactoring) the session creation and ending of selenium into a new file called browser.py

Hope that helps.

@sionking Nice one, I think the only things that are left are the ones @uluQulu commented (especially the self argument in one of the methods). Other than that the changes look good. 
I haven't tested the PR locally but I'll do that soon.

**Review by timgrossmann — APPROVED** [review]

**Comment by the-unknown** [comment]
Will this, in the end, lead to an InstaPy that doesn't need to have a browser installed?
So is this the first step into a future completely without selenium?

**Comment by timgrossmann** [comment]
@the-unknown Well, (un)fortunately not. We've had some internal discussions of removing the selenium part completely and only using requests without any browser installation needed.

The thing is, that having this "human"-like automated browser interactions is what separates InstaPy from any other tool out there.
We would need much more research to being able to actively say that we can completely automate Instagram without an active browser session the same way we are doing it with selenium.

Also, having selenium in there fights the "commercial"-overuse of this tool since you can't simply run 1000 sessions in parallel like you could with a non browser based approach.

Another thing is the possibility to create more "human"-like features like mouse movement, random clicking, scrolling, story views etc. Instagram records everything you are doing, so in the long run we think that having an automation like InstaPy in place will be more sustainable than using plain requests.

Hope you can understand that.

Edit: Sorry, so to actually answer your question. This won't help with this, but it will get rid of many selenium (browser) related issues in the future.

**Comment by Mehran** [comment]
@timgrossmann so Clear explain, thanks.

**Comment by timgrossmann** [comment]
@sionking Sorry for the conflict that resolved from merging the PR from @uluQulu 

Could you please rebase your PR before we merge it?

File: instapy/browser.py [hunk 4]
```diff
@@ -130,16 +132,15 @@ def set_selenium_local_session(proxy_address,
 
     browser.implicitly_wait(page_delay)
 
-    # TODO: need to move to called
+    #
```
Comment (converge) [inline instapy/browser.py]: dont leave me alone in the dark 😄

File: instapy/browser.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        matches = re.match(r'^(\d+\.\d+)',
+                           browser.capabilities['chrome']['chromedriverVersion'])
+        if float(matches.groups()[0]) < Settings.chromedriver_min_version:
+            err_msg = 'chromedriver {} is not supported, expects {}+'.format(
+                            float(matches.groups()[0]), Settings.chromedriver_min_version)
+            return browser, err_msg
+
+    browser.implicitly_wait(page_delay)
+
+    #
```
Comment (timgrossmann) [inline instapy/browser.py]:
```suggestion
```

**Comment by sionking** [comment]
> @sionking Sorry for the conflict that resolved from merging the PR from @uluQulu
> 
> Could you please rebase your PR before we merge it?

done

**Comment by timgrossmann** [comment]
@sionking Awesome, thank you very much!

Let's merge this as the first step towards a better and less error prune browser session #3581

**Pull request merged**